### PR TITLE
DecompositionBase

### PR DESCRIPTION
**Pull request by Beakerboy**
Collecting some of the common functions together in one place. Reducing lines of code, increasing coverage.

### CODE REVIEW

File: src/LinearAlgebra/Decomposition/DecompositionBase.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace MathPHP\LinearAlgebra\Decomposition;
+
+use MathPHP\Exception;
+use MathPHP\LinearAlgebra\Matrix;
+
+abstract class DecompositionBase implements \ArrayAccess
```
Comment (markrogoyski) [inline src/LinearAlgebra/Decomposition/DecompositionBase.php]: How about just naming it `Decomposition` without the Base suffix?

**Comment by Beakerboy** [comment]
Done. Let me know if there’s anything else.

File: src/LinearAlgebra/Decomposition/Decomposition.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+use MathPHP\Exception;
+use MathPHP\LinearAlgebra\Matrix;
+
+abstract class Decomposition implements \ArrayAccess
+{
+    abstract public static function decompose(Matrix $M);
+
+    /**************************************************************************
+     * ArrayAccess INTERFACE
```
Comment (markrogoyski) [inline src/LinearAlgebra/Decomposition/Decomposition.php]: How about also adding `offsetExists` as an abstract empty method that each Decomposition has to implement.

Comment (Beakerboy) [reply src/LinearAlgebra/Decomposition/Decomposition.php]: Good idea

**Comment by markrogoyski** [comment]
Looks good. Thanks for continuing to improve the Matrix classes.

**Pull request merged**